### PR TITLE
ENH: Add PDF, CDF and parameter estimation for Stable Distributions

### PR DESCRIPTION
**Pull request by bsdz**
Add Levy Stable Parameter Estimation using McCulloch 1986 Quantiles method.
Add PDF and CDF calculation using Zolotarev's method and FFT estimate on Continuous Fourier
Integral of Characteristic Function.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: `x, alpha, beta = np.broadcast_arrays(x, alpha, beta)` ?

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Is this correct? Isnt' the result from fitstart an approximation (it uses interpolation tables).

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Prefer splrep + BSpline over *UnivariateSpline.

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Take real part before interpolating?

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: `x, alpha, beta = np.broadcast_arrays(x, alpha, beta)` instead?

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Reference format to use:
```
[MC] McCulloch, J., 1986. Simple consistent estimators of stable distribution parameters.
     Communications in Statistics - Simulation and Computation 15, 11091136.
[MS] Mittnik, S.T. rachev, T. Doganoglu, D. Chenyao, 1999. Maximum likelihood estimation 
     of stable Paretian models,  Mathematical and Computer Modelling, Volume 29, Issue 10, 1999, Pages 275-293,
```
Also note that there appear to be some typos here.

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Fitting uses the quantile estimation method in [MC]_.

File: scipy/stats/_continuous_distns.py
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Is it correct to return constant results here?

**Review by pv — COMMENTED** [review]
Some general comments. Not reviewing the algorithm/implementation itself.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: You are right that fitstart is an approximation and the quantile estimate works well here. However, the MLE method breaks for other reasons so I thought it best to override that so at least one can at least estimate parameters.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: I believe this is correct. Similar to other distributions, ie:

semicircular_gen._stats
uniform_gen._stats
vonmises_gen._stats
wald_gen._stats

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: This needs to match the spline as obtained by interp1d. Would the two you suggest do that?

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: Changed - thanks.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: Changed - thanks.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: Changed - thanks.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: Changed - thanks.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [inline scipy/stats/_continuous_distns.py]: Yes makes sense - updated. Thanks.

File: scipy/stats/_continuous_distns.py
Comment (an81) [reply scipy/stats/_continuous_distns.py]: It is really exciting to see stable laws implemented in the scipy. I have spend few years working with different density implementations. There are asymptotic expansions and intergral represenations, which performs pretty well and are more stable numerically-wise then fft. Also there are different parametrizations of the complex parameter in the exponent of characteristic function, which I believe are more intuitive. I guess it could be next development.

**Comment by bsdz** [comment]
Hi @an81, sorry in advance if I have misinterpreted your question. I think if all goes well, and this PR passes review, it will be included for release 1.1.

**Comment by an81** [comment]
Sure, well, I have spend few years on working on stable laws, written paper
about them, wrote a code for densities ...
Its not the best idea to use fft, because its numerically unstable.
I can add the code and we see if it passes.

**Comment by an81** [comment]
So on my behalf, I honestly hope that if all you use is naive fft,
that the code will not pass, because it will create tons of problems if
somebody would relly on the implementation
and blindly used the function without testing it.

**Comment by bsdz** [comment]
hi Andrea, thanks for your suggestions. I'm not sure if you've looked at the code in detail but it actually uses FFT or quad. It depends on how many points are requested. This is included in the documentation I provided. I don't  think FFT is as bad as you seem to suggest, provided you choose a sufficient number of points. Please do follow the references provided in the code. Also it might be worth reading Stable Paretian Models in Finance by Rachev/Mittnik where they provide a fairly comprehensive coverage of the topic (albeit a little outdated). I've been using a similar FFT approach in various models since 2008 in a commercial setting and a similar approach is used in other popular financial products. Also please do submit code changes and references as I'm sure this PR could be improved.  It's certainly not perfect.

**Comment by an81** [comment]
I dont like the work of Rachev & Mittnik.
I know their work, other people did better job.
We also looked for the stable for the industry (finance) purpose ...
My main criticism for fft is the Gibbs effect, i.e. the tails fluctuates.

I guess there is work of Nolan, who literally just re-written Zolotarev's
papers.
Its easy to workout he integral representation, as Zolotarev did ( and
which Nolan had implemented end of 90??? )
Anyway, I think it could be more helpful to just share the work which is
done on this topic.
Indeed there expansions which works pretty well for certain regions.

As you know, you can represent densities also as special functions (Fox's
or Meier G- function)
so trying this cn get better results and testing environment too.

To be fair, I really wonder why the people in finance dont use more of
stable laws (tempered stable laws).
Maybe because few of them tried fft and realized that the Gibbs effect is
not realiable ...

**Comment by bsdz** [comment]
@an81, I feel the biggest challenge in finance is more often simplicity, in the sense: can you explain it to someone who doesn't have a phd in maths; and also speed, in implementation and calculation. FFT allows you produce a near complete density in one go allowing you to interpolate several points at once. Perhaps this might be possible with special functions too. Tbh I don't know much about representing densities with Fox or Meier G functions. I'll be happy to take a look at these methods and implement them into this PR as ideally it would be nice to remove/reduce the Gibbs effect. Can you point to any papers with derivation and/or implementation details?

**Comment by an81** [comment]
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.117.6505&rep=rep1&type=pdf

**Comment by bsdz** [comment]
Hi @an81. Thank you for the 550+ page book. Please can you be a bit more specific? Some sample code goes a long way too. Also can you perhaps test the existing code and highlight where the Gibbs effect might be more prominent? eg low alpha etc; perhaps this can be just documented with a recommendation that users use quad in these cases (already in code). This is until better implementation is available.

**Comment by an81** [comment]
Blair, is there a way to have a chat via email?

**Comment by bsdz** [comment]
I've added Zolotarev's method for PDF calculation as suggested. Now we support 3 methods. The newer method passes same tests as existing quadrature method, ie matches Nolan's model results. There's still the Zolotarev's CDF method to implement although this shouldn't be too tricky. Until then the CDF method still relies on numerical integration of the PDF using the FFT method.

**Comment by bsdz** [comment]
I've also added Zolotarev's method for CDF calculation. Since Zolotarev CDF method is much more accurate, the general CDF calculation will no longer use interpolated spline integrals on FFT PDF by default. The machinery is still there for later revival if one wishes to increase performance. Also, increased minimum point threshold for FFT PDF to 100 points.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [reply scipy/stats/_continuous_distns.py]: @pv I removed the fit() override so this now uses MLE again. using interpolation tables as initial estimate. I simply reordered the test cases and it works; this could be a memory corruption somewhere. I suspect it's in the core library; if I can pin it down I'll raise a separate issue.

**Review by mikofski — COMMENTED** [review]
Thank you so much for your contribution. I am very impressed!

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [reply scipy/stats/_continuous_distns.py]: Thanks! :-) 

Updated docstrings as suggested.

**Review by an81 — CHANGES_REQUESTED** [review]
Add doc_string on levy_stable dist (line: 3581)

File: scipy/stats/tests/test_distributions.py
Comment (rgommers) [inline scipy/stats/tests/test_distributions.py]: This causes the one test failure in the last CI run. Use `suppress_warnings` instead (check example from another test if you search for that name).

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [inline scipy/stats/_continuous_distns.py]: This warning doesn't quite match what you said in https://github.com/scipy/scipy/pull/8766#issuecomment-396062917. Can this be more precise (e.g. `in particular when |alpha-1| < 0.005; alpha < 0.2 or alpha=1 and beta is very small `)

**Comment by rgommers** [comment]
Code style looks fine and CI checkers are happy. There's a lot of too long lines, but let's leave those in to not make rebasing gh-8766 on top of this one too difficult.

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [reply scipy/stats/_continuous_distns.py]: also, would be good to use:

    .. warning::

        text about alpha ranges

so this actually gets noticed by users.

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [inline scipy/stats/_continuous_distns.py]: This whole branch doesn't have test coverage (see [coverage report](https://codecov.io/gh/scipy/scipy/compare/b582245fe89dd7363f650b3da8124929b39ce13a...910a63f0de8704f1d95c0ef565823e2c21b4c737/src/scipy/stats/_continuous_distns.py?before=scipy/stats/_continuous_distns.py)), could you try to construct a test that exercises this code?

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [inline scipy/stats/_continuous_distns.py]: This branch is also missing test coverage.

File: scipy/stats/_continuous_distns.py
Comment (bsdz) [reply scipy/stats/_continuous_distns.py]: Just regarding my previous comment, that only applies to the output from Nolan's stablec and may or may not apply to our methods. Since his executable is known to be unstable for alpha < 0.2 we don't extract data for that range. Specifically, we now extract densities for alpha = .25, .5, .75, 1, 1.25, 1.5 and 2, beta = -1, -.5, 0, .5, 1 and x = -10..10 (step 1) and check each method against subsets of these ranges up to specific decimal precision. Hope that sounds reasonable.

**Comment by rgommers** [comment]
This now has several test failures (https://travis-ci.org/scipy/scipy/jobs/391535771) and a number of PEP8 issues (https://travis-ci.org/scipy/scipy/jobs/391535766)

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [reply scipy/stats/_continuous_distns.py]: Okay, thanks for the clarification

**Comment by bsdz** [comment]
@rgommers I've fixed the code style and most test failures. Although, it appears [for the osx (darwin) build server some of the tests fail with some calculations returning a lower precision](https://travis-ci.org/scipy/scipy/builds/391686367). I'm not sure how to deal with this as I don't have access to that platform. One idea is to separate the tests by platform, eg:

```python
tests = [
    ...
    # zolatarev is accurate except at alpha==1
    ['zolotarev', None, 8, lambda r: (sys.platform != 'darwin') & (r['alpha'] != 1)],
    ['zolotarev', None, 6, lambda r: (sys.platform == 'darwin') & (r['alpha'] != 1)],
    ...
]
```

Although I'd have to test what precision will pass on osx through trial and error by commits to this PR and travis build triggers. Does that sound reasonable or perhaps there's a better way?

**Comment by rgommers** [comment]
> Although I'd have to test what precision will pass on osx through trial and error by commits to this PR and travis build triggers. Does that sound reasonable or perhaps there's a better way?

I have a macOS machine, I can give it a try. CI is too slow to do trial and error without it getting annoying.

**Comment by rgommers** [comment]
> One idea is to separate the tests by platform, eg:

That's only a good idea if it's clear what in a given platform causes a specific problem. In this case I don't see a reason why macOS should be worse.

The test is not fixable by changing the test precision - the function return has `nan` in it. The problem seems to be in these lines:

        with np.errstate(all="ignore"):
            intg_max = optimize.minimize_scalar(lambda theta: -f(theta), bounds=[-xi, np.pi/2])
            intg = integrate.quad(f, -xi, np.pi/2, points=[intg_max.x])[0]

**Comment by ilayn** [comment]
Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

**Comment by bsdz** [comment]
@ilayn Thanks for the heads up. I'll work from a separate branch as suggested and merge once resolved.

@rgommers It appears the integration is failing rather than the minimization. I'll see if I can figure out what parameter values are problematic.

**Comment by bsdz** [comment]
@rgommers I've fixed the MacOS issue and all tests pass now. It seems that quad() behaviour differs slightly between Linux, MacOS and Windows with Linux being the most forgiving. For example, on linux if we pass in a point outside the integration bounds it will gracefully continue but on Windows and MacOS it will fail. This happens because minimize_scalar() doesn't guarantee returning values within requested bounds. Also Linux will be happy if we pass bounds that are the same (null point integral) and just return zero but this will fail on other platforms. This could be down to floating point differences and in one case I use isclose() to check if the endpoints match as "==" doesn't work on MacOS.

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [inline scipy/stats/_continuous_distns.py]: `isclose` has a default relative tolerance of 1e-5. I think what you intended here is to check for equality with close to machine precision? this throws away too much accuracy it looks like.

**Comment by rgommers** [comment]
> This could be down to floating point differences and in one case I use isclose() to check if the endpoints match as "==" doesn't work on MacOS.

In general, you can never rely on exact equality for floating point numbers. You can even get things like `0` and `-0` not comparing equal. You normally want to check with `isclose` or some such function indeed, but with a relative/absolute tolerance that's appropriate (typically 1e-15 or so for float64).

**Comment by rgommers** [comment]
There's 6 merges of scipy master into this branch. Typically we want to avoid that unless absolutely necessary, because it makes the history more messy. In this case let's leave it to not make the rebases in the other PR harder.

**Comment by rgommers** [comment]
These two runtime warnings still appear (at least on macOS):

```
stats/tests/test_distributions.py::TestLevyStable::()::test_pdf_nolan_samples
  /Users/rgommers/Code/scipy/scipy/stats/_continuous_distns.py:3780: RuntimeWarning: Density calculation unstable for alpha=1 and beta!=0. Use quadrature method instead.
    ' Use quadrature method instead.', RuntimeWarning)
  /Users/rgommers/Code/scipy/scipy/stats/_continuous_distns.py:3880: RuntimeWarning: Density calculations experimental for FFT method. Use combination of zolatarev and quadrature methods instead.
    ' Use combination of zolatarev and quadrature methods instead.', RuntimeWarning)

stats/tests/test_distributions.py::TestLevyStable::()::test_cdf_nolan_samples
  /Users/rgommers/Code/scipy/scipy/stats/_continuous_distns.py:3917: RuntimeWarning: Cumulative density calculations experimental for FFT method. Use zolatarev method instead.
    ' Use zolatarev method instead.', RuntimeWarning)
```

They should be filtered out inside those tests by using

    with suppress_warnings() as sup:
        sup.record(RuntimeWarning, "<start of the warning message>*")

**Comment by rgommers** [comment]
almost there @bsdz :)

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [reply scipy/stats/_continuous_distns.py]: covered now

File: scipy/stats/_continuous_distns.py
Comment (rgommers) [reply scipy/stats/_continuous_distns.py]: covered now

**Comment by rgommers** [comment]
All right, in it goes. Thanks @bsdz, nice improvement! And thanks @an81, @mikofski, @pv for the reviews.

Further improvement of these methods in gh-8766.

**Pull request merged**
